test/unit/index/fm_index_iterator_test.cpp

```diff
@@ -0,0 +1,234 @@
+// ============================================================================
+//                 SeqAn - The Library for Sequence Analysis
+// ============================================================================
+//
+// Copyright (c) 2006-2017, Knut Reinert & Freie Universitaet Berlin
+// Copyright (c) 2016-2017, Knut Reinert & MPI Molekulare Genetik
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+//     * Redistributions of source code must retain the above copyright
+//       notice, this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above copyright
+//       notice, this list of conditions and the following disclaimer in the
+//       documentation and/or other materials provided with the distribution.
+//     * Neither the name of Knut Reinert or the FU Berlin nor the names of
+//       its contributors may be used to endorse or promote products derived
+//       from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL KNUT REINERT OR THE FU BERLIN BE LIABLE
+// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+// LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+// OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+// DAMAGE.
+//
+// ============================================================================
+
+#include <type_traits>
+
+#include <gtest/gtest.h>
+
+#include <seqan3/index/fm_index.hpp>
+
+#include <seqan3/alphabet/nucleotide/dna4.hpp>
```

Comment:
I think, we should test for all the alphabets we support?
The to_rank() functionality is already tested in the alphabet tests, but the fm_index still has stuff that is otherwise untested, i.e. [`sdsl::int_vector<8> tmp_text(text.size());`](https://github.com/seqan/seqan3/blob/65f041ed72568e9b852915f1260671fd1ae7efe4/include/seqan3/index/fm_index.hpp#L161), which should produce errors once we use UTF alphabets and have a character with rank > 255.

Reply:
as mentioned before, we shouldn't support larger alphabets than for example amino acid alphabets. in that case the user should use indices for integer alphabets (where I currently don't see any application for in sequence analysis)

but we might enforce using byte-alphabets. any idea how to do that? we can't check for the underlying type (::char_type) since indices should work with types such as char, uint8_t as well

Reply:
I think you can use `std::is_same_v<underlying_rank<t>::type, uint8_t>`, which should be implemented for all types t that satisfy the alphabet_concept.
It's [here ](https://github.com/seqan/seqan3/blob/master/include/seqan3/alphabet/detail/member_exposure.hpp#L69) by and included with `#include <seqan3/alphabet/concept.hpp>`.